File: Periodo4/CG/TP2/code/classes/camera.py
import numpy as np

from numpy import array as Vec3
from numpy.random import uniform as rand

from classes.ray import Ray

from functions.vector import normalize

# Função auxiliar que retorna um ponto aleatório em um disco (círculo)
def random_point_in_disk():
    r = np.sqrt(rand())
    theta = 2 * np.pi * rand()

    return Vec3([r * np.cos(theta), r * np.sin(theta), .0])

class Camera:
    """
        Classe para representar nossa camera
    """

    # Método para inicializar uma classe do tipo Camera
    def __init__(self, origin, look_at, vup, fov, aspect, aperture, focus_dist):
        theta = np.radians(fov)
        half_height = np.tan(theta / 2)
        half_width = aspect * half_height

        self.__w = normalize(origin - look_at)
        self.__u = normalize(np.cross(vup, self.__w))
        self.__v = np.cross(self.__w, self.__u)
        self.__len_radius = aperture / 2

        self.__origin = origin
        self.__lower_left_corner = origin - focus_dist * (self.__u * half_width + self.__v * half_height + self.__w)
        self.__horizontal = 2 * self.__u * focus_dist * half_width
        self.__vertical = 2 * self.__v * focus_dist * half_height

    # Método para retornar a posição da câmera
    def origin(self):
        return self.__origin

    # Método para dispararmos um raio
    def shoot_ray(self, s, t):
        rd = self.__len_radius * random_point_in_disk()
        offset = self.__u * rd[0] + self.__v * rd[1]

        origin = self.origin() + offset
        direction = self.__lower_left_corner + s * self.__horizontal + t * self.__vertical - self.__origin - offset
        return Ray(origin, direction)

File: ComputacaoGrafica/RayTracer/src/camera.cpp
#include "camera.h"

// Definindo o construtor da classe
Camera::Camera(const Vec3& origin, const Vec3& look_at, const Vec3& vup, const float fov,
               const float aspect, const float aperture, const float focus_dist) {

    float theta = fov * M_PI / 180;
    float half_height = tan(theta / 2);
    float half_width = aspect * half_height;

    this->w = Vec3::normalize(origin - look_at);
    this->u = Vec3::normalize(Vec3::cross(vup, w));
    this->v = Vec3::cross(w, u);

    this->len_radius = aperture / 2;
    this->lower_left_corner = origin - focus_dist * (half_width * u + half_height * v + w);
    this->horizontal = 2 * half_width * focus_dist * u;
    this->vertical = 2 * half_height * focus_dist * v;
    this->origin = origin;
}

// Definindo o destrutor da classe
Camera::~Camera() {}

// Definindo método que retorna um ponto aleatório em um círculo unitário
Vec3 Camera::random_in_unit_disk() const {
    float r = sqrt(drand48());
    float t = 2 * M_PI * drand48();

    float x = r * cos(t);
    float y = r * sin(t);

    return Vec3(x, y, 0);
}

// Definindo o método que dispara raios
Ray Camera::shoot_ray(const float s, const float t) const {
    Vec3 rd = this->len_radius * random_in_unit_disk();
    Vec3 offset = u * rd.get_x() + v * rd.get_y();
    Vec3 dir = this->lower_left_corner + s * this->horizontal + t * this->vertical - this->origin;
    return Ray(this->origin + offset, dir - offset);
}